### PR TITLE
Add SLES support

### PR DESCRIPTION
**Pull request by aiyengar2**
This commit PR adds support for versioned docker installs in SLES

Related Issue: https://github.com/rancher/rancher/issues/25342

### CODE REVIEW

**Comment by aiyengar2** [comment]
Tested this against SLES 12 SP 2 and SLES 12 SP 5 for x86_64 while providing different docker versions by changing the docker_version value.

This commit should be able to be replicated across older docker versions as it:
1) tries to resolve the requested version of docker
2) tries to resolve for any docker version with the same major minor version (i.e. trunc_docker_version)
3) Fails out if no matching major minor version can be found in zypper

File: 19.03.5.sh [hunk 1]
```diff
@@ -332,6 +335,10 @@ do_install() {
 			dist_version="$(rpm -q --whatprovides ${lsb_dist}-release --queryformat "%{VERSION}\n" | sed 's/\/.*//' | sed 's/\..*//' | sed 's/Server*//' | sort | tail -1)"
 		;;
 
+		sles)
+			dist_version="$(rpm -q --whatprovides ${lsb_dist}-release --queryformat "%{VERSION}\n" | sed 's/\/.*//' | sort | tail -1)"
+		;;
+
```
Comment (aiyengar2) [inline 19.03.5.sh]: This value of dist_version is not being used right now by SLES docker install but I left it there because:
1) It would allow the ability to quickly fail out based on the dist_version value if it lacks support (i.e. see centos|fedora|redhat|oraclelinux case)
2) It is consistent with the other OSs offered.

Please let me know if you have any thoughts!

File: 19.03.5.sh [hunk 2]
```diff
@@ -511,6 +518,36 @@ do_install() {
...
+		sles)
+			(
+			set -x
+			$sh_c 'zypper refresh'
+			docker_pkg=$(zypper se -n -s docker | grep -E "\sdocker\s.*${docker_version//\./\\\.}.*$architecture" | sort --sort=v -k3 | tail -1 | cut -d '|' -f4 | cut -d ' ' -f2)
+			if [ -z "$docker_pkg" ]; then
+			    cat >&2 <<-'EOF'
+				No exact match found for SLES that matches docker \$docker_version for \$architecture.
+
+				Attempting to install the latest available version of docker \${docker_version%.*}.
```
Comment (superseb) [inline 19.03.5.sh]: Was this a specific requirement for SLES? The script installs the requested version (`19.03.5` in this case) or fails. Getting a different version from the one you requested is unexpected.

**Comment by superseb** [comment]
Is there a dependency on a preconfigured repository or do we need to add it to the system?

**Comment by aiyengar2** [comment]
> Is there a dependency on a preconfigured repository or do we need to add it to the system?

We install docker through the repository SLE-Module-Containers<SUSE Version / SP>-Updates (i.e. for SLES 15 SP2 this comes from SLE-Module-Containers15-SP1-Updates). 

This should be available as a module (Containers) on any member of the SUSE Linux Enterprise Product Family according to: https://documentation.suse.com/sles/15-SP1/pdf/art-modules_color_en.pdf

Either way, do you think it might make sense to add an additional check to the script to ensure the repository exists? If it does not, we shouldn't be able to search (i.e. `zypper se`) for the docker package anyways so this might be a redundant check, but we could output a different log accordingly if that is the case.

File: 19.03.5.sh (hunk 2 above)
Comment (aiyengar2) [reply 19.03.5.sh]: This should only affect the patch requested but will still fail if the major / minor version requested is incorrect i.e. since SUSE Linux only has 19.03.1 and 19.03.5, so if we ported these changes to the 19.03.4 or 19.03.6 (if/when it releases) install-docker script, the current implementation would still automatically pick up 19.03.5.

I do agree that getting a different version that then one requested is unexpected since, if a user did not have a preference for the patch they would like to use, they would have just decided to use the major.minor version of the script (i.e. 19.03), which is already symbolically linked to the latest version. However, in the case that we roll out a new version of install docker (i.e. 19.03.6) and the newest version does not have SLES support, then an existing user using SLES who has always requested for 19.03 (which now points to 19.03.6) would now suddenly be unable to install docker on their machines post the commit to point 19.03 -> 19.03.6 (since there is no support for 19.03.6 in SLES in this hypothetical situation).

Therefore, I think that either we can:
1) remove this logic and request users using SLES to point to a specific docker version / patch that is available on SLES rather than just 19.03, in which case they would be responsible for upgrading to the latest versions of docker on SLES when we add support.
2) retain this logic with the expectation that, for SLES, we perform a "best effort" installation to either get the patch requested or the latest patch adhering to that same major/minor version

@deniseschannon thoughts?

**Comment by aiyengar2** [comment]
Versions available on SLES 12:

```
ec2-user@ip-<@aiyengar2: redacted>:~> zypper se -n -s docker | grep -E "\sdocker\s" | sort --sort=v -k3
v | docker                          | package    | 1.6.2-31.2                            | x86_64 | SLE-Module-Containers12-Pool     
v | docker                          | package    | 1.6.2-37.3                            | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 1.7.1-40.1                            | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 1.8.1-43.3                            | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 1.8.2-46.1                            | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 1.8.3-49.1                            | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 1.8.3-55.2                            | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 1.9.1-58.1                            | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 1.10.3-63.1                           | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 1.10.3-66.1                           | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 1.10.3-71.1                           | x86_64 | SLE-Module-Containers12-Updates  
i | docker                          | package    | 1.11.2-76.1                           | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 1.12.3-81.2                           | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 1.12.5-84.1                           | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 1.12.6-87.2                           | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 1.12.6-90.1                           | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 17.04.0_ce-98.2                       | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 17.09.1_ce-98.8.1                     | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 17.09.1_ce-98.12.1                    | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 17.09.1_ce-98.15.1                    | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 17.09.1_ce-98.18.1                    | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 18.06.1_ce-98.21.1                    | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 18.06.1_ce-98.24.1                    | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 18.09.1_ce-98.34.2                    | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 18.09.6_ce-98.37.1                    | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 18.09.6_ce-98.40.1                    | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 18.09.7_ce-98.43.1                    | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 19.03.1_ce-98.46.1                    | x86_64 | SLE-Module-Containers12-Updates  
v | docker                          | package    | 19.03.5_ce-98.51.1                    | x86_64 | SLE-Module-Containers12-Updates  
```

Versions available on SLES 15:
```
ec2-user@ip-<@aiyengar2: redacted>:~> zypper se -n -s docker | grep -E "\sdocker\s" | sort --sort=v -k3
v  | docker                          | package    | 18.09.1_ce-6.14.1                     | x86_64 | SLE-Module-Containers15-SP1-Pool     
v  | docker                          | package    | 18.09.6_ce-6.17.1                     | x86_64 | SLE-Module-Containers15-SP1-Updates  
v  | docker                          | package    | 18.09.6_ce-6.20.3                     | x86_64 | SLE-Module-Containers15-SP1-Updates  
v  | docker                          | package    | 18.09.6_ce-6.23.1                     | x86_64 | SLE-Module-Containers15-SP1-Updates  
i+ | docker                          | package    | 19.03.1_ce-6.26.2                     | x86_64 | SLE-Module-Containers15-SP1-Updates  
v  | docker                          | package    | 19.03.5_ce-6.31.1                     | x86_64 | SLE-Module-Containers15-SP1-Updates  
   | docker                          | srcpackage | 18.09.6_ce-6.17.1                     | noarch | SLE-Module-Containers15-SP1-Updates  
   | docker                          | srcpackage | 18.09.6_ce-6.20.3                     | noarch | SLE-Module-Containers15-SP1-Updates  
   | docker                          | srcpackage | 18.09.6_ce-6.23.1                     | noarch | SLE-Module-Containers15-SP1-Updates  
   | docker                          | srcpackage | 19.03.1_ce-6.26.2                     | noarch | SLE-Module-Containers15-SP1-Updates  
   | docker                          | srcpackage | 19.03.5_ce-6.31.1                     | noarch | SLE-Module-Containers15-SP1-Updates  
```

**Comment by aiyengar2** [comment]
Discussed offline:
- In the case that an SLES version is unavailable, the expected behavior should be to fail to install docker and print an error statement that lists the available docker versions within the requested $major.$minor version (i.e. a failure to install 19.03.6 should print that only 19.03.1 and 19.03.5 are available in SLES)
- Customers using SLES should use a specific $major.$minor.$patch install-docker script instead of the symbolically linked $major.$minor install-docker script as the latest version of docker available may not be supported by SLES.

File: 19.03.5.sh (hunk 1 above)
Comment (aiyengar2) [reply 19.03.5.sh]: As of the latest commit, this value is now being used.

File: 19.03.5.sh (hunk 2 above)
Comment (aiyengar2) [reply 19.03.5.sh]: discussed offline (described in comments)

**Comment by aiyengar2** [comment]
> Discussed offline:
> 
> * In the case that an SLES version is unavailable, the expected behavior should be to fail to install docker and print an error statement that lists the available docker versions within the requested $major.$minor version (i.e. a failure to install 19.03.6 should print that only 19.03.1 and 19.03.5 are available in SLES)
> * Customers using SLES should use a specific $major.$minor.$patch install-docker script instead of the symbolically linked $major.$minor install-docker script as the latest version of docker available may not be supported by SLES.

Latest commit has been updated to follow this behavior.

File: 19.03.5.sh [hunk 3]
```diff
@@ -284,6 +284,9 @@ do_install() {
 	if [ -z "$lsb_dist" ] && [ -r /etc/redhat-release ]; then
 		lsb_dist='redhat'
 	fi
+	if [ -z "$lsb_dist" ] && [ -r /etc/SuSE-release ]; then
```
Comment (superseb) [inline 19.03.5.sh]: We can probably drop this, as this is deprecated and might be removed (even in Service Pack):

```
# cat /etc/SuSE-release
SUSE Linux Enterprise Server 12 (x86_64)
VERSION = 12
PATCHLEVEL = 5
# This file is deprecated and will be removed in a future service pack or release.
# Please check /etc/os-release for details about this release.
```

The fallback to `os-release` should work.

File: 19.03.5.sh [hunk 4]
```diff
@@ -511,6 +518,32 @@ do_install() {
...
 			exit 0
 			;;
+		sles)
+			(
+			set -x
+			$sh_c 'zypper refresh'
+			docker_pkg=$(zypper se -n -s docker | grep -E "\sdocker\s.*${docker_version//\./\\.}.*$architecture" | sort --sort=v -k3 | tail -1 | cut -d '|' -f4 | cut -d ' ' -f2)
+			if [ -z "$docker_pkg" ]; then
+			    trunc_docker_version=${docker_version%.*}
+				available_docker_pkgs=$(zypper se -n -s docker | grep -E "\sdocker\s.*${trunc_docker_version//\./\\.}.*$architecture" | sort --sort=v -k3 | cut -d '|' -f4 | cut -d ' ' -f2 | cut -d '_' -f1)
```
Comment (superseb) [inline 19.03.5.sh]: This seems indented for no reason.

File: 19.03.5.sh [hunk 5]
```diff
@@ -511,6 +518,32 @@ do_install() {
...
+				cat >&2 <<-EOF
+				No match found for SLES that matches docker $docker_version for $architecture.
```
Comment (superseb) [inline 19.03.5.sh]: Can we check if anything was found, the output on systems that have no Docker available at all:

```
No match found for SLES that matches docker 19.03.5 for x86_64.
                                                                                                                                                                                                                                                                                                                         
SLES 12.5 only supports the following patches within docker 19.03: 

```

Can we also use `Docker` as its a product name

File: 19.03.5.sh [hunk 6]
```diff
@@ -511,6 +518,32 @@ do_install() {
...
 			;;
+		sles)
+			(
+			set -x
+			$sh_c 'zypper refresh'
+			docker_pkg=$(zypper se -n -s docker | grep -E "\sdocker\s.*${docker_version//\./\\.}.*$architecture" | sort --sort=v -k3 | tail -1 | cut -d '|' -f4 | cut -d ' ' -f2)
+			if [ -z "$docker_pkg" ]; then
+			    trunc_docker_version=${docker_version%.*}
+				available_docker_pkgs=$(zypper se -n -s docker | grep -E "\sdocker\s.*${trunc_docker_version//\./\\.}.*$architecture" | sort --sort=v -k3 | cut -d '|' -f4 | cut -d ' ' -f2 | cut -d '_' -f1)
+				cat >&2 <<-EOF
```
Comment (superseb) [inline 19.03.5.sh]: Can we rename this to (if this makes sense)

```
No package for docker found that matches requested version $docker_version for $architecture

Packages found for docker ${trunc_docker_version}:
```

File: 19.03.5.sh [hunk 7]
```diff
@@ -511,6 +518,32 @@ do_install() {
 			)
 			exit 0
 			;;
+		sles)
+			(
+			set -x
+			$sh_c 'zypper refresh'
+			docker_pkg=$(zypper se -n -s docker | grep -E "\sdocker\s.*${docker_version//\./\\.}.*$architecture" | sort --sort=v -k3 | tail -1 | cut -d '|' -f4 | cut -d ' ' -f2)
+			if [ -z "$docker_pkg" ]; then
+			    trunc_docker_version=${docker_version%.*}
```
Comment (superseb) [inline 19.03.5.sh]: Does it make sense to use the included `semverParse()` for this or not really?

File: 19.03.5.sh (hunk 4 above)
Comment (aiyengar2) [reply 19.03.5.sh]: Updated in latest commit!

File: 19.03.5.sh (hunk 6 above)
Comment (aiyengar2) [reply 19.03.5.sh]: Updated in latest commit!

File: 19.03.5.sh (hunk 3 above)
Comment (aiyengar2) [reply 19.03.5.sh]: Updated in latest commit!

File: 19.03.5.sh (hunk 5 above)
Comment (aiyengar2) [reply 19.03.5.sh]: Added this to the latest commit. On failure, it would output:
`SLES 12.5 does not support any patches for docker 19.03 for x86_64.`

File: 19.03.5.sh (hunk 7 above)
Comment (aiyengar2) [reply 19.03.5.sh]: I utilized `semverParse` in the latest commit, but either approach should be okay! semverParse pulls out $major, $minor, and $patch, which we then concatenate to $major.$minor, whereas the previous substitution just removes $patch from docker_version.

File: 19.03.5.sh [hunk 8]
```diff
@@ -511,6 +515,39 @@ do_install() {
...
+			set -x
+			$sh_c 'zypper refresh'
+			docker_pkg=$(zypper se -n -s docker | grep -E "\sdocker\s.*${docker_version//\./\\.}.*$architecture" | sort --sort=v -k3 | tail -1 | cut -d '|' -f4 | cut -d ' ' -f2)
+			if [ -z "$docker_pkg" ]; then
+				semverParse $docker_version
+				trunc_docker_version=$major.$minor
+				available_docker_pkgs=$(zypper se -n -s docker | grep -E "\sdocker\s.*${trunc_docker_version//\./\\.}.*$architecture" | sort --sort=v -k3 | cut -d '|' -f4 | cut -d ' ' -f2 | cut -d '_' -f1)
+				if [ -z "$available_docker_pkgs" ]; then
+					cat >&2 <<-EOF
+					SLES $dist_version does not support any patches for docker $trunc_docker_version for $architecture.
```
Comment (superseb) [inline 19.03.5.sh]: Not sure why I missed this before, this implies the OS does not support it but this is not necessarily true as the repo can be missing to provide the docker package. We should just tell the user there was no package found for Docker, and to check enabled repositories.

**Comment by superseb** [comment]
Do we need `-q` for quiet like used in the other package managers?

**Comment by aiyengar2** [comment]
> Do we need `-q` for quiet like used in the other package managers?

Good catch, added it to the latest commit!

File: 19.03.5.sh (hunk 8 above)
Comment (aiyengar2) [reply 19.03.5.sh]: Updated in the latest commit!

Comment (aiyengar2) [reply 19.03.5.sh]: Updated in the latest commit!

**Comment by aiyengar2** [comment]
> Only situation where the script can hang is when the repository's signing key is not accepted, we have to double check if this situation can be avoided by just logging the error that key is not accepted and not hang:
> 
> ```
> New repository or package signing key received:
>   Repository:       Virtualization:containers (SLE_12_SP4)                         
>   Key Name:         Virtualization OBS Project <Virtualization@build.opensuse.org>
>   Key Fingerprint:  55A0B34D 49501BB7 CA474F5A A193FBB5 72174FC2                   
>   Key Created:      Mon Feb 17 05:10:11 2020                                                                                 
>   Key Expires:      Wed Apr 27 05:10:11 2022                                       
>   Rpm Name:         gpg-pubkey-72174fc2-5e4a2033                                   
> Do you want to reject the key, trust temporarily, or trust always? [r/t/a/? shows all options] (r):
> ```

Addressed in latest commit! Added in the `-n` flag to `zypper refresh`

**Comment by aiyengar2** [comment]
Regarding the addition of the `-n` (non-interactive) flag to `zypper refresh`:

If `zypper -q -n refresh` encounters any issues that require user intervention (such as untrusted keys belonging to any added repository, as referenced in the comments above), the install-docker script will fail. 

@superseb and I discussed offline about how this poses an issue to a user who installs an unrelated repository using `zypper addrepo <repo>` and then tries to run this script. This user would encounter an error on running the script, even if the requested patch of docker exists in a repository that was able to be refreshed. (i.e. in the example below, `docker 19.03.5` exists in the `SLE-Module-Containers15-SP1-Updates` repo but the install still fails due to an issue with refreshing `Virtualization:containers`)

However, we ultimately decided against changing this behavior as failing to refresh all zypper repositories could result in an unintentional security bug. For example, let's say that a user intended to use a more updated version of `docker 19.03.5` in `Virtualization:containers` instead of the one found in `SLE-Module-Containers15-SP1-Updates`; if we were to not fail on refresh, the user would potentially unintentionally be using an older version of the package. Between the two options, it's preferable to go with the false negative (a failure to install with a clear error statement to the user on how to resolve the issue) over a false positive.

TLDR - As a result of this change, the install-docker script relies on being able to successfully run `zypper refresh` on a user's machine before installing docker.

Example Failure:
```
ec2-user@ip-172-31-35-135:~> ./install-docker.sh 
Warning: the "docker" command appears to already exist on this system.
If you already have Docker installed, this script can cause trouble, which is
why we're displaying this warning and provide the opportunity to cancel the
installation.
If you installed the current Docker package using this script and are using it
again to update Docker, you can safely ignore this message.
You may press Ctrl+C now to abort this script.
+ sleep 20
+ sudo -E sh -c 'zypper -n -q refresh'
New repository or package signing key received:
  Repository:       Virtualization:containers (SLE_15_SP1)                        
  Key Name:         Virtualization OBS Project <Virtualization@build.opensuse.org>
  Key Fingerprint:  55A0B34D 49501BB7 CA474F5A A193FBB5 72174FC2                  
  Key Created:      Mon 17 Feb 2020 05:10:11 AM UTC                               
  Key Expires:      Wed 27 Apr 2022 05:10:11 AM UTC                               
  Rpm Name:         gpg-pubkey-72174fc2-5e4a2033                                  
Do you want to reject the key, trust temporarily, or trust always? [r/t/a/?] (r): r
Repository 'Virtualization:containers (SLE_15_SP1)' is invalid.
[Virtualization_containers|http://download.opensuse.org/repositories/Virtualization:/containers/SLE_15_SP1/] Valid metadata not found at specified URL
Skipping repository 'Virtualization:containers (SLE_15_SP1)' because of the above error.
Some of the repositories have not been refreshed because of an error.
```
